repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -77,82 +87,83 @@ class Session(
-    }
+      statement.compareAndTransit(StatementState.Running, StatementState.Available)
+      statement.compareAndTransit(StatementState.Cancelling, StatementState.Cancelled)
+    }(interpreterExecutor)
 
     statementId
   }
 
   def cancel(statementId: Int): Unit = {
-    if (!_statements.contains(statementId)) {
+    val statementOpt = _statements.synchronized { _statements.get(statementId) }
+    if (statementOpt.isEmpty) {
       return
     }
 
-    if (_statements(statementId).state.get() == StatementState.Available ||
-      _statements(statementId).state.get() == StatementState.Cancelled ||
-      _statements(statementId).state.get() == StatementState.Cancelling) {
+    val statement = statementOpt.get
+    if (statement.state.get().isOneOf(
+      StatementState.Available, StatementState.Cancelled, StatementState.Cancelling)) {
       return
     } else {
       // statement 1 is running and statement 2 is waiting. User cancels
       // statement 2 then cancels statement 1. The 2nd cancel call will loop and block the 1st
       // cancel call since cancelExecutor is single threaded. To avoid this, set the statement
       // state to cancelled when cancelling a waiting statement.
-      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Cancelled)
-      _statements(statementId).state.compareAndSet(
-        StatementState.Running, StatementState.Cancelling)
+      statement.compareAndTransit(StatementState.Waiting, StatementState.Cancelled)
+      statement.compareAndTransit(StatementState.Running, StatementState.Cancelling)
     }
 
     info(s"Cancelling statement $statementId...")
 
-      Future {
-        val deadline = livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TIMEOUT).millis.fromNow
-        while (_statements(statementId).state.get() == StatementState.Cancelling) {
-          if (deadline.isOverdue()) {
-            info(s"Failed to cancel statement $statementId.")
-            _statements(statementId).state.compareAndSet(
-              StatementState.Cancelling, StatementState.Cancelled)
-          } else {
-            _sc.foreach(_.cancelJobGroup(statementId.toString))
-          }
-          Thread.sleep(livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TRIGGER_INTERVAL))
-        }
-        if (_statements(statementId).state.get() == StatementState.Cancelled) {
-          info(s"Statement $statementId cancelled.")
+    Future {
+      val deadline = livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TIMEOUT).millis.fromNow
+
+      while (statement.state.get() == StatementState.Cancelling) {
+        if (deadline.isOverdue()) {
+          info(s"Failed to cancel statement $statementId.")
+          statement.compareAndTransit(StatementState.Cancelling, StatementState.Cancelled)
+        } else {
+          _sc.foreach(_.cancelJobGroup(statementId.toString))
         }
-      }(cancelExecutor)
+        Thread.sleep(livyConf.getTimeAsMs(RSCConf.Entry.JOB_CANCEL_TRIGGER_INTERVAL))
```

Comment:
LGTM. One last thing, avoid sleeping if statement state has changed to Cancelled. Will merge after fixing this.